Classes/Promise.swift

```diff
@@ -0,0 +1,269 @@
+            if let queue = queue {
+                queue.async {
+                    self.internalCall(with: object)
+                }
+            } else {
+                internalCall(with: object)
+            }
+        }
+    }
+    
+    @discardableResult func then(_ onFulfilled: ((ValueType) -> Void)? = nil,
+                                 observed: ((ValueType) -> Void)? = nil,
+                                 rejected : ((Error) -> Void)? = nil,
+                                 always : (() -> Void)? = nil,
+                                 queue : DispatchQueue) -> PromiseCancelToken {
+        let cancellationToken = PromiseCancelToken()
+        let blocks = PromiseBlock(cancellationToken: cancellationToken)
+        blocks.queue = queue
+        blocks.done = onFulfilled
+        blocks.observed = observed
+        blocks.rejected = rejected
+        blocks.always = always
+        
+        cancellationToken.cancelled({ [weak self, unowned blocks] in
+            Promise.queue.async {
+                if let `self` = self {                    
+                    blocks.done = nil
+                    blocks.observed = nil
+                    blocks.rejected = nil
+                    blocks.always = nil
+                    
+                    var strongCount = false
+                    for block in `self`.blocks {
+                        if block.shouldKeepPromise {
+                            strongCount = true
+                            break
+                        }
+                    }
+                    if strongCount {
+                        `self`.cancel()
+                    }
+                }
+            }
+        }, on: queue)
+
+        if let fulfilledObject = self.fulfilledObject, fulfilledObject.resolved {
+            blocks.call(with: fulfilledObject)
+        } else {
+            Promise.queue.async {
+                if let fulfilledObject = self.fulfilledObject, fulfilledObject.resolved {
+                    blocks.call(with: fulfilledObject)
+                } else {
+                    self.blocks.append(blocks)
+                }
+            }
+        }
+        
+        return cancellationToken
+    }
+    
+    lazy var blocks: Array<PromiseBlock> = []
+
+    override func cancelled(_ onCancel: @escaping () -> Void, on queue: DispatchQueue) {
+        let wrappedBlock = {
+            queue.async {
+                onCancel()
+            }
+        }
+        
+        PromiseCancelToken.queue.async {
+            guard let fulfilledObject = self.fulfilledObject, fulfilledObject.resolved else {
+                if self.cancelled {
+                    wrappedBlock()
+                } else {
+                    self.onCancel = wrappedBlock
+                }
+                return
+            }
+        }
+    }
+
+}
+
+// fulfillment
+extension Promise {
+    
+    func fulfill(with: PromiseResult<ValueType>) {
```

Comment:
about name of func, in my view, your variant func `call(with object: PromiseResult<ValueType>) {` with the object above is better
` func fulfill(with result: PromiseResult<ValueType>)`